experimental/stats/metricregistry.go

```diff
@@ -0,0 +1,240 @@
+// registered with the same name, this function will panic.
+func RegisterFloat64Count(descriptor MetricDescriptor) Float64CountHandle {
+	registerMetric(descriptor.Name, descriptor.Default)
+	descriptor.Type = MetricTypeFloatCount
+	handle := Float64CountHandle{
+		MetricDescriptor: &descriptor,
+	}
+	MetricsRegistry[descriptor.Name] = &descriptor
+	return handle
+}
+
+// RegisterInt64Histo registers the metric description onto the global registry.
+// It returns a typed handle to use to recording data.
+//
+// NOTE: this function must only be called during initialization time (i.e. in
+// an init() function), and is not thread-safe. If multiple metrics are
+// registered with the same name, this function will panic.
+func RegisterInt64Histo(descriptor MetricDescriptor) Int64HistoHandle {
+	registerMetric(descriptor.Name, descriptor.Default)
+	descriptor.Type = MetricTypeIntHisto
+	handle := Int64HistoHandle{
+		MetricDescriptor: &descriptor,
+	}
+	MetricsRegistry[descriptor.Name] = &descriptor
+	return handle
+}
+
+// RegisterFloat64Histo registers the metric description onto the global
+// registry. It returns a typed handle to use to recording data.
+//
+// NOTE: this function must only be called during initialization time (i.e. in
+// an init() function), and is not thread-safe. If multiple metrics are
+// registered with the same name, this function will panic.
+func RegisterFloat64Histo(descriptor MetricDescriptor) Float64HistoHandle {
+	registerMetric(descriptor.Name, descriptor.Default)
+	descriptor.Type = MetricTypeFloatHisto
+	handle := Float64HistoHandle{
+		MetricDescriptor: &descriptor,
+	}
+	MetricsRegistry[descriptor.Name] = &descriptor
+	return handle
+}
+
+// RegisterInt64Gauge registers the metric description onto the global registry.
+// It returns a typed handle to use to recording data.
+//
+// NOTE: this function must only be called during initialization time (i.e. in
+// an init() function), and is not thread-safe. If multiple metrics are
+// registered with the same name, this function will panic.
+func RegisterInt64Gauge(descriptor MetricDescriptor) Int64GaugeHandle {
+	registerMetric(descriptor.Name, descriptor.Default)
+	descriptor.Type = MetricTypeIntGauge
+	handle := Int64GaugeHandle{
+		MetricDescriptor: &descriptor,
+	}
+	MetricsRegistry[descriptor.Name] = &descriptor
+	return handle
+}
+
+// MetricType is the type of metric.
+type MetricType int
+
+const (
+	MetricTypeIntCount MetricType = iota
+	MetricTypeFloatCount
+	MetricTypeIntHisto
+	MetricTypeFloatHisto
+	MetricTypeIntGauge
+)
+
+// clearMetricsRegistryForTesting clears the global data of the metrics
+// registry. It returns a closure to be invoked that sets the metrics registry
+// to its original state. Only called in testing functions.
+func clearMetricsRegistryForTesting() func() {
+	oldDefaultMetrics := DefaultMetrics
+	oldRegisteredMetrics := registeredMetrics
+	oldMetricsRegistry := MetricsRegistry
```

Comment:
This shouldn't clear.  It should observe how things are before the test runs only, then reset the state to that when the test is done.

Also can we pass `testing.T` here and use `t.Cleanup`?

Reply:
Discussed offline; tests should be robust against prepopulated state in the instrument registry as that simulates real world scenario, clear only if absolutely needed.